Environment variable assignments on Configure command line

levitte opened the pull request:
For Unix, it's pretty easy to temporarly assign environment variables on the command line:

    CC=clang ./config

On other platforms, it's not as easy, so therefore, this implements an alternative (VMS shown as example):

    perl Configure "CC=whatevercc" vms-alpha

Futhermore, this also save away the environment values more effectively, and also makes sure the saved away values are used before anything on the environment when reconfiguring.

levitte committed:
Save away the environment variables we rely on

There are cases when we overwrite %ENV values, and while this is
perfectly fine on some platforms, it isn't on others, because the
Configure script isn't necessarely run in a separate process, and
thus, changing %ENV may very well change the environment of the
calling shell.  VMS is such a platform.

Furthermore, saving away values that we use also allow us to save them
in configdata.pm in an effective way, and recall those values just as
effectively when reconfiguring.  Also, this makes sure that we do use
the saved away values when reconfiguring, when the actual environment
variables might otherwise affect us.

levitte committed:
Make it possible to add env var assignments as Configure options

In other words, make the following possible:

    ./config CC=clang

or

    ./Configure CC=clang linux-x86_64

levitte committed:
Document the possibility for command line argument env assignments

richsalz commented:
Should we use a flag like ```-e FOO=...``` rather than just matching on the pattern?

levitte commented:
Why?  This is already a well known pattern, both with `make` and `autoconf``s `configure`...  and it doesn't compete with current option patterns either.

levitte committed:
Have all relevant config targets use the env() function rather than $ENV

This way, any of the relevant environment variables for the platform
being configured are preserved and don't have to be recalled manually
when reconfiguring.

richsalz commented:
Okay, fine, thanks for the answer.

dot-asm commented on Configure:
```diff
@@ -2522,6 +2534,15 @@ sub which
+
+    return $config{perlenv}->{$name} if exists $config{perlenv}->{$name};
+    $config{perlenv}->{$name} = $ENV{$name};
+    return $config{perlenv}->{$name};
```
Wouldn't
```
$config{perlenv}->{$name} = $ENV{$name} if ! exists $confg{peleenv}->{$name};
return $config{perlenv}->{$name};
```
be more transparent?

levitte replied:
Good point.  Will fix

dot-asm commented on Configure:
```diff
@@ -223,25 +223,13 @@ if (grep /^reconf(igure)?$/, @argvcopy) {
-	$ENV{$local_config_envname} = $configdata::config{local_config_dir}
-	    if defined($configdata::config{local_config_dir});
+	$config{perlenv} = { defined($configdata::config{perlenv}) ?
+	    %{$configdata::config{perlenv}} : () };
```
Why does one need outermost {}?

levitte replied:
I could do `$config{perlenv} = $configdata::config{perlenv}`, it's true...  experience has given me reason to be paranoid and to want to do a shallow copy of the hash rather than just copying the reference.

dot-asm replied:
I'm not really fond of this, because it's too subtle. I mean ideally code should not evoke more questions than necessary, especially such esoteric. So a commentary?

levitte replied:
I added documentation in `INSTALL`, does that work for you?

dot-asm replied:
I meant specifically subtle use of outer {} *here*. Comment elsewhere doesn't really cut it, because it's as problematic to make the connection to specific language syntax as unravel why it's used :-)

levitte replied:
I decided to be less paranoid instead ;-)

dot-asm commented on Configure:
```diff
@@ -551,6 +539,14 @@ my @seed_sources = ();
 	$_ = shift @argvcopy;
+
+	# Support env variable assignments among the options
+	if (m|^(\w+)=(.+)?$|)
```
Someone will want specify something with spaces, in which case there probably should be a way to specify that...

dot-asm replied:
Do you know what people will complain about? "Why can't I specify CLFAGS? That's how we do it elsewhere..." Well, CFLAGS environment variable is ignored now, answer is pass on command line, say with `./config $CFLAGS`. But question here is if one decides to do something, would it be appropriate to do something different. BTW, CFLAGS might be too loaded to think about, consider rather LDFLAGS as think experiment...

levitte replied:
I'm not sure what you mean with "specify something with spaces"...  wouldn't something like this suffice?

```
./config SPACE=' '
```

Yeah, I know about the CFLAGS / LDFLAGS demand, and am currently having another look.  It should be fairly easy if we do it the same way as the GNU folks, i.e. support CFLAGS, CPPFLAGS and LDFLAGS.

dot-asm replied:
> I'm not sure what you mean ...

Environment variable with space in its value. `SPACE=' '` is kind of bad example, because it kind of works out now. But try `SPACE='a b'`...

dot-asm replied:
Of course there is difference between passing SPACE='a b' to ./config and ./Configure... If passed to ./Configure it works as expected. So there was a bit of confusion there, sorry...

levitte replied:
Ahhhhhh, `config` doesn't quotify the arguments to `Configure`.... ok, that should be easy enough to fix, I hope

dot-asm replied:
Why `(.+)?` and not `(.*)`?

levitte replied:
Because I want an empty value (no characters at all) to generate `undef`

levitte replied:
`config` to `Configure` argument passing fixed, or so I hope

dot-asm commented:
> would it be appropriate to do something different.

Speaking of something different. As *side* note. It might be useful to "freeze" some variables. Well, it's not something one can do by assigning variables in makefile, but one might do something through template... Just in case for reference, In VC case it might be interesting to "freeze" environment variables such as `CL`, `_CL_`, `INCLUDE`, `LIBPATH`... So that one can re-run make "in another window" without having to replicate environment... Well, just a wild thought, might be worth considering in "background"...

levitte commented:
> Speaking of something different. As *side* note. It might be useful to "freeze" some variables. Well, it's not something one can do by assigning variables in makefile, but one might do something through template... Just in case for reference, In VC case it might be interesting to "freeze" environment variables such as `CL`, `_CL_`, `INCLUDE`, `LIBPATH`... So that one can re-run make "in another window" without having to replicate environment... Well, just a wild thought, might be worth considering in "background"...

Uhmmm, so preserving the environment in `{perlenv}` and recalling it when a `./Configure reconf` is made, doesn't that do exactly what you ask for?  Or oh, are those environment variables that we ignore but that are used by the compiler?  That *could* be done, and quite easily so, just add this in your favorite targets:

``` perl
        env => [ env('CL'), env('_CL_'), ... ],
```

That attribute is unused, but will automatically add those variables to `$config{perlenv}`...

...  but that makes me think that we probably need a lazy variant of `sub env`...

dot-asm commented:
> Uhmmm, so preserving the environment in {perlenv} and recalling it when a ./Configure reconf is made, doesn't that do exactly what you ask for?

Once again, it's just a wild thought that *might* be worth considering, *eventually* [or never]. Being a "wild thought" ii might be inconsistent and turn out dumb. But to elaborate. This is not only about environment variables at config time, but even at build time. This is why I brought up makefile template...

levitte commented:
I see. I did think about that, but didn't finish the thought...

levitte committed:
Make sure ./config passes options to ./Configure correctly

This is, even when they contain spaces or all kinds of funny quotes

dot-asm commented on Configure:
```diff
@@ -2522,6 +2539,22 @@ sub which
+        return $config{perlenv}->{$name};
+    };
+
+    return $reading_targets ? $fn : $fn->();
```
The problem I have with this approach is that it's another detail/dependency you have to keep in mind/maintain. In other words I'd say that *less* "strange" dependencies is better. [Which by the way the idea behind "target factory" I've discussed in private in #4684 context, it's an attempt to unify member instantiation without member-specific hooks.] Also note that most invocations of env are already lazy. Because most of them appear in lazily-called subs. So wouldn't it be more appropriate to say that if any of remaining calls, i.e. those not currently lazy, has to be lazy for some reason, then it could be made lazy [by placing it into anonymous sub]?

levitte replied:
Ahhh, good point.  Okie, will redo this.

dot-asm commented:
I wonder what's intention with squashing? I mean will everything be squashed?

levitte commented:
I don't generally squash everything into a single commit.  That's simply not how I do things.  The exception is when I explicitely use `git commit --fixup` or `git commit --squash`, those commits will be automatically treated by a final `git rebase --autosquash`.

levitte committed:
Configure: die if there are other arguments with 'reconf'

It's better to inform the user about this than silently ignoring
something that the user might expect to work, somehow.

levitte committed:
Document how the configuration option 'reconf' works

dot-asm commented on INSTALL:
```diff
@@ -532,6 +532,30 @@
+                   Reconfigure from earlier data.  This fetches the previous
+                   command line options and environment from data saved in
+                   "configdata.pm", and runs the configuration process again,
+                   uring these options and environment.
```
"using"

levitte replied:
Fixed

dot-asm commented:
> I don't generally squash everything into a single commit.

I wasn't directly suggesting to, but at the same time I see no point in keeping small adjustments resulting from feedback.

levitte commented:
> I wasn't directly suggesting to, but at the same time I see no point in keeping small adjustments resulting from feedback.

I'll have another look...  but I try to partition into pieces that hold their own, so to say.

levitte committed:
fixup! Save away the environment variables we rely on

levitte commented:
Ok, I *think* I've adressed all concerns expressed so far.

levitte commented:
Merged.  7ecdf18d80cba14ad1afa6c0d18574d2ad2929c3 to e7a206694451be19432d079691610994473f53b7

levitte commented:
Thank you!
